File: device/metal/attention/native/scaled_dot_product.m
#include "scaled_dot_product.h"
#include "attention.h"
#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>

static int metal_attention_softmax_pipeline(
    MetalContext* context,
    MetalStatus* status,
    id<MTLComputePipelineState>* pipeline
) {
    *pipeline = metal_get_pipeline(context, "attention_softmax", status);

    if (*pipeline == nil) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    return 0;
}

static int metal_attention_prepare(
    MetalContext* context,
    int elementDType,
    MetalStatus* status,
    id<MTLComputePipelineState>* scoresPipeline,
    id<MTLComputePipelineState>* softmaxPipeline,
    id<MTLComputePipelineState>* weightedPipeline
) {
    int scoresCode = metal_attention_pipeline(
        context, "attention_scores", elementDType, status, scoresPipeline
    );

    if (scoresCode != 0) {
        return scoresCode;
    }

    int softmaxCode = metal_attention_softmax_pipeline(context, status, softmaxPipeline);

    if (softmaxCode != 0) {
        return softmaxCode;
    }

    return metal_attention_pipeline(
        context, "attention_weighted", elementDType, status, weightedPipeline
    );
}

static int metal_attention_encode_scores(
    id<MTLCommandBuffer> commandBuffer,
    id<MTLComputePipelineState> pipeline,
    MetalStatus* status,
    MetalBufferRef queryRef,
    MetalBufferRef keyRef,
    MetalBufferRef scoresRef,
    uint32_t seqQ,
    uint32_t seqK,
    uint32_t depth
) {
    id<MTLComputeCommandEncoder> encoder =
        metal_attention_encoder(commandBuffer, pipeline, status);

    if (encoder == nil) {
        return status != NULL && status->code != 0 ? status->code : -4;
    }

    [encoder setBuffer:(__bridge id<MTLBuffer>)queryRef offset:0 atIndex:0];
    [encoder setBuffer:(__bridge id<MTLBuffer>)keyRef offset:0 atIndex:1];
    [encoder setBuffer:(__bridge id<MTLBuffer>)scoresRef offset:0 atIndex:2];
    [encoder setBytes:&seqQ length:sizeof(seqQ) atIndex:3];
    [encoder setBytes:&seqK length:sizeof(seqK) atIndex:4];
    [encoder setBytes:&depth length:sizeof(depth) atIndex:5];
    [encoder
        dispatchThreadgroups:MTLSizeMake((seqK + 15) / 16, (seqQ + 15) / 16, 1)
        threadsPerThreadgroup:MTLSizeMake(16, 16, 1)
    ];
    [encoder endEncoding];
    return 0;
}

static int metal_attention_encode_softmax(
    id<MTLCommandBuffer> commandBuffer,
    id<MTLComputePipelineState> pipeline,
    MetalStatus* status,
    MetalBufferRef scoresRef,
    uint32_t seqQ,
    uint32_t seqK
) {
    id<MTLComputeCommandEncoder> encoder =
        metal_attention_encoder(commandBuffer, pipeline, status);

    if (encoder == nil) {
        return status != NULL && status->code != 0 ? status->code : -4;
    }

    [encoder setBuffer:(__bridge id<MTLBuffer>)scoresRef offset:0 atIndex:0];
    [encoder setBytes:&seqK length:sizeof(seqK) atIndex:1];
    [encoder
        dispatchThreadgroups:MTLSizeMake(seqQ, 1, 1)
        threadsPerThreadgroup:MTLSizeMake(256, 1, 1)
    ];
    [encoder endEncoding];
    return 0;
}

static int metal_attention_encode_weighted(
    id<MTLCommandBuffer> commandBuffer,
    id<MTLComputePipelineState> pipeline,
    MetalStatus* status,
    MetalBufferRef valueRef,
    MetalBufferRef scoresRef,
    MetalBufferRef outRef,
    uint32_t seqQ,
    uint32_t seqK,
    uint32_t valueDim
) {
    id<MTLComputeCommandEncoder> encoder =
        metal_attention_encoder(commandBuffer, pipeline, status);

    if (encoder == nil) {
        return status != NULL && status->code != 0 ? status->code : -4;
    }

    [encoder setBuffer:(__bridge id<MTLBuffer>)scoresRef offset:0 atIndex:0];
    [encoder setBuffer:(__bridge id<MTLBuffer>)valueRef offset:0 atIndex:1];
    [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:2];
    [encoder setBytes:&seqQ length:sizeof(seqQ) atIndex:3];
    [encoder setBytes:&seqK length:sizeof(seqK) atIndex:4];
    [encoder setBytes:&valueDim length:sizeof(valueDim) atIndex:5];
    [encoder
        dispatchThreadgroups:MTLSizeMake((valueDim + 15) / 16, (seqQ + 15) / 16, 1)
        threadsPerThreadgroup:MTLSizeMake(16, 16, 1)
    ];
    [encoder endEncoding];
    return 0;
}

int metal_dispatch_attention(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef queryRef,
    MetalBufferRef keyRef,
    MetalBufferRef valueRef,
    MetalBufferRef scoresRef,
    MetalBufferRef outRef,
    uint32_t seqQ,
    uint32_t seqK,
    uint32_t depth,
    uint32_t valueDim,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        metal_attention_status_clear(status);

        if (queryRef == NULL || keyRef == NULL || valueRef == NULL ||
            scoresRef == NULL || outRef == NULL) {
            metal_transformer_status_set(status, -2, "nil Metal buffer");
            return -2;
        }

        MetalContext* context = (MetalContext*)contextRef;
        id<MTLComputePipelineState> scoresPipeline = nil;
        id<MTLComputePipelineState> softmaxPipeline = nil;
        id<MTLComputePipelineState> weightedPipeline = nil;
        int prepareCode = metal_attention_prepare(
            context,
            elementDType,
            status,
            &scoresPipeline,
            &softmaxPipeline,
            &weightedPipeline
        );

        if (prepareCode != 0) {
            return prepareCode;
        }

        id<MTLCommandBuffer> commandBuffer = metal_attention_command_buffer(context, status);

        if (commandBuffer == nil) {
            return status != NULL && status->code != 0 ? status->code : -3;
        }

        int scoresCode = metal_attention_encode_scores(
            commandBuffer, scoresPipeline, status,
            queryRef, keyRef, scoresRef,
            seqQ, seqK, depth
        );

        if (scoresCode != 0) {
            return scoresCode;
        }

        int softmaxCode = metal_attention_encode_softmax(
            commandBuffer, softmaxPipeline, status, scoresRef, seqQ, seqK
        );

        if (softmaxCode != 0) {
            return softmaxCode;
        }

        int weightedCode = metal_attention_encode_weighted(
            commandBuffer, weightedPipeline, status,
            valueRef, scoresRef, outRef,
            seqQ, seqK, valueDim
        );

        if (weightedCode != 0) {
            return weightedCode;
        }

        metal_track_command_completion((MetalContext*)contextRef, commandBuffer, completionToken, NULL);
        [commandBuffer commit];
        return 0;
    }
}

File: device/cuda/attention/native/scaled_dot_product.c
#include "scaled_dot_product.h"
#include "attention.h"
#include "../internal/bridge/core_private.h"

static int cuda_attention_get_kernel(
    CUDAContext* context,
    const char* operationName,
    int elementDType,
    CUDAKernelRef* kernel,
    CUDAStatus* status
) {
    char kernelName[128];
    int nameCode = cuda_attention_kernel_name(
        kernelName,
        sizeof(kernelName),
        operationName,
        elementDType,
        status
    );

    if (nameCode != 0) {
        return nameCode;
    }

    const char* moduleSource = cuda_attention_module_source();

    if (moduleSource == NULL) {
        cuda_attention_status_clear(status);
        cuda_transformer_status_set(status, -7, "CUDA attention module source not registered");
        return -7;
    }

    *kernel = cuda_get_kernel(context, moduleSource, kernelName, status);

    if (*kernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    return 0;
}

static int cuda_attention_launch_scores(
    CUDAContext* context,
    CUDAKernelRef kernel,
    CUDAStreamRef stream,
    CUDABufferRef queryRef,
    CUDABufferRef keyRef,
    CUDABufferRef scoresRef,
    uint32_t seqQ,
    uint32_t seqK,
    uint32_t depth,
    CUDAStatus* status
) {
    void* queryPtr = cuda_buffer_device_ptr(queryRef);
    void* keyPtr = cuda_buffer_device_ptr(keyRef);
    void* scoresPtr = cuda_buffer_device_ptr(scoresRef);
    void* args[] = {&queryPtr, &keyPtr, &scoresPtr, &seqQ, &seqK, &depth};

    return cuda_launch_grid(
        context,
        kernel,
        stream,
        (seqK + 15u) / 16u,
        (seqQ + 15u) / 16u,
        1u,
        16u,
        16u,
        1u,
        0u,
        args,
        sizeof(args),
        status
    );
}

static int cuda_attention_launch_softmax(
    CUDAContext* context,
    CUDAKernelRef kernel,
    CUDAStreamRef stream,
    CUDABufferRef scoresRef,
    uint32_t seqQ,
    uint32_t seqK,
    CUDAStatus* status
) {
    void* scoresPtr = cuda_buffer_device_ptr(scoresRef);
    void* args[] = {&scoresPtr, &seqK, &seqQ};

    return cuda_launch_grid(
        context,
        kernel,
        stream,
        seqQ,
        1u,
        1u,
        256u,
        1u,
        1u,
        0u,
        args,
        sizeof(args),
        status
    );
}

static int cuda_attention_launch_weighted(
    CUDAContext* context,
    CUDAKernelRef kernel,
    CUDAStreamRef stream,
    CUDABufferRef scoresRef,
    CUDABufferRef valueRef,
    CUDABufferRef outRef,
    uint32_t seqQ,
    uint32_t seqK,
    uint32_t valueDim,
    CUDAStatus* status
) {
    void* scoresPtr = cuda_buffer_device_ptr(scoresRef);
    void* valuePtr = cuda_buffer_device_ptr(valueRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&scoresPtr, &valuePtr, &outPtr, &seqQ, &seqK, &valueDim};

    return cuda_launch_grid(
        context,
        kernel,
        stream,
        (valueDim + 15u) / 16u,
        (seqQ + 15u) / 16u,
        1u,
        16u,
        16u,
        1u,
        0u,
        args,
        sizeof(args),
        status
    );
}

int cuda_dispatch_attention(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef queryRef,
    CUDABufferRef keyRef,
    CUDABufferRef valueRef,
    CUDABufferRef scoresRef,
    CUDABufferRef outRef,
    uint32_t seqQ,
    uint32_t seqK,
    uint32_t depth,
    uint32_t valueDim,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_attention_status_clear(status);

    if (queryRef == NULL || keyRef == NULL || valueRef == NULL ||
        scoresRef == NULL || outRef == NULL) {
        cuda_transformer_status_set(status, -2, "nil CUDA attention buffer");
        return -2;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    int prepareCode = cuda_context_prepare(contextRef, status, &context, &stream);

    if (prepareCode != 0) {
        return prepareCode;
    }

    CUDAKernelRef scoresKernel = NULL;
    int scoresPrepareCode = cuda_attention_get_kernel(
        context, "attention_scores", elementDType, &scoresKernel, status
    );

    if (scoresPrepareCode != 0) {
        return scoresPrepareCode;
    }

    CUDAKernelRef softmaxKernel = NULL;
    const char* moduleSource = cuda_attention_module_source();

    if (moduleSource == NULL) {
        cuda_transformer_status_set(status, -7, "CUDA attention module source not registered");
        return -7;
    }

    softmaxKernel = cuda_get_kernel(context, moduleSource, "attention_softmax", status);

    if (softmaxKernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    CUDAKernelRef weightedKernel = NULL;
    int weightedPrepareCode = cuda_attention_get_kernel(
        context, "attention_weighted", elementDType, &weightedKernel, status
    );

    if (weightedPrepareCode != 0) {
        return weightedPrepareCode;
    }

    int scoresCode = cuda_attention_launch_scores(
        context, scoresKernel, stream,
        queryRef, keyRef, scoresRef,
        seqQ, seqK, depth, status
    );

    if (scoresCode != 0) {
        return scoresCode;
    }

    int softmaxCode = cuda_attention_launch_softmax(
        context, softmaxKernel, stream, scoresRef, seqQ, seqK, status
    );

    if (softmaxCode != 0) {
        return softmaxCode;
    }

    int weightedCode = cuda_attention_launch_weighted(
        context, weightedKernel, stream,
        scoresRef, valueRef, outRef,
        seqQ, seqK, valueDim, status
    );

    if (weightedCode != 0) {
        return weightedCode;
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}
